sound/soc/soc-core.c

```diff
@@ -2831,10 +2831,22 @@ EXPORT_SYMBOL_GPL(snd_soc_register_card);
 
 static void snd_soc_unbind_card(struct snd_soc_card *card, bool unregister)
 {
+	struct snd_soc_pcm_runtime *rtd;
+	int order;
+
 	if (card->instantiated) {
 		card->instantiated = false;
 		snd_soc_dapm_shutdown(card);
 		snd_soc_flush_all_delayed_work(card);
+
+		/* remove all components used by DAI links on this card */
+		for_each_comp_order(order) {
+			for_each_card_rtds(card, rtd) {
+				soc_remove_link_components(card, rtd, order);
```

Comment:
this patch is the one I am most scared of. I don't see a moral equivalent in snd_soc_bind_card that would add link components and don't know the internals of the ASoC code enough to judge if this is legit or not. Either you find a self-explanatory way of doing this or you have to run this by Mark Brown.

Reply:
@plbossart let me submit a RFC patchset to the mailing list to get some feedback. What I did find was as far as I could tell none of the component drivers implement the remove() callback. So I doubt this will break anything. But I'll check with him anyway.

Reply:
@plbossart that was easy. the most controversial one was the first one to be accepted. I'll clean up this PR based on Mark's feedback. I'll drop the first one and keep the rest.